theano/sandbox/cuda/cula.py

```
@@ -0,0 +1,129 @@
+import theano
+from theano.sandbox.cuda.type import CudaNdarrayType
+from theano.sandbox.cuda import GpuOp
+
+from theano.sandbox.cuda.basic_ops import as_cuda_ndarray_variable
+from theano.sandbox.cuda import cuda_ndarray
+
+dimshuffle = cuda_ndarray.cuda_ndarray.dimshuffle
+
+cula_available = False
+
+try:
+    from scikits.cuda import cula
+    cula_available = True
+except (ImportError, OSError):
+    pass
+
+cula_initialized = False
+
+
+class GpuSolve(GpuOp):
+    """
+    CULA GPU solver OP.
+
+    :param trans: Whether to take the transpose of the input matrix
+    or not.
+    """
+    __props__ = ('trans',)
+
+    def __init__(self, trans='N'):
+        self.trans = trans
+        super(GpuSolve, self).__init__()
+
+    def output_type(self, inp):
+        return CudaNdarrayType(broadcastable=[False] * inp.type.ndim)
+
+    def make_node(self, inp1, inp2):
+        inp1 = as_cuda_ndarray_variable(inp1)
+        inp2 = as_cuda_ndarray_variable(inp2)
+
+        assert inp1.ndim == 2
+        assert inp2.ndim == 2
+        return theano.Apply(self, [inp1, inp2], [self.output_type(inp1)()])
+
+    def make_thunk(self,
+                   node,
+                   storage_map, _,
+                   no_recycling=[]):
+
+        # Initialize CULA the first time it is needed
+        global cula_initialized
+        if cula_available and cula and not cula_initialized:
```

if cula_available is false, we should raise an error here.